File: project/drivers/driver/php/src/LeakDetector.php
<?php
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

/**
 * ScratchBird PHP Driver - Connection Leak Detector
 * Copyright (c) 2025-2026 Dalton Calford
 */

namespace ScratchBird;

enum LeakLogLevel: string {
    case DEBUG = 'debug';
    case WARN = 'warn';
    case ERROR = 'error';
}

class LeakDetectionConfig {
    public int $thresholdMs = 30000;
    public bool $captureStackTrace = false;
    public int $checkIntervalMs = 10000;
    public LeakLogLevel $logLevel = LeakLogLevel::WARN;
    
    public function __construct(array $options = []) {
        $this->thresholdMs = $options['thresholdMs'] ?? 30000;
        $this->captureStackTrace = $options['captureStackTrace'] ?? false;
        $this->checkIntervalMs = $options['checkIntervalMs'] ?? 10000;
        $this->logLevel = $options['logLevel'] ?? LeakLogLevel::WARN;
    }
}

class CheckoutInfo {
    public float $checkoutTime;
    public int $threadId;
    public ?string $stackTrace;
    public array $metadata;
    
    public function __construct(bool $captureStackTrace, array $metadata = []) {
        $this->checkoutTime = microtime(true);
        $this->threadId = getmypid();
        $this->metadata = $metadata;
        $this->stackTrace = $captureStackTrace ? debug_backtrace() : null;
    }
    
    public function getHeldDurationMs(): int {
        return (int)((microtime(true) - $this->checkoutTime) * 1000);
    }
}

class LeakDetectionGuard {
    private LeakDetector $detector;
    private string $connectionId;
    private bool $released = false;
    
    public function __construct(LeakDetector $detector, string $connectionId) {
        $this->detector = $detector;
        $this->connectionId = $connectionId;
    }
    
    public function release(): void {
        if (!$this->released) {
            $this->detector->checkin($this->connectionId);
            $this->released = true;
        }
    }
    
    public function __destruct() {
        $this->release();
    }
}

class LeakDetector {
    private LeakDetectionConfig $config;
    private array $checkouts = [];
    private bool $running = false;
    
    public function __construct(LeakDetectionConfig $config = null) {
        $this->config = $config ?? new LeakDetectionConfig();
    }
    
    public function start(): void {
        $this->running = true;
    }
    
    public function stop(): void {
        $this->running = false;
    }
    
    public function checkout(string $connectionId, array $metadata = []): LeakDetectionGuard {
        $this->checkouts[$connectionId] = new CheckoutInfo($this->config->captureStackTrace, $metadata);
        return new LeakDetectionGuard($this, $connectionId);
    }
    
    public function checkin(string $connectionId): void {
        $info = $this->checkouts[$connectionId] ?? null;
        if ($info) {
            unset($this->checkouts[$connectionId]);
            if ($info->getHeldDurationMs() > $this->config->thresholdMs) {
                // Log held too long
            }
        }
    }
    
    public function getActiveCount(): int {
        return count($this->checkouts);
    }
    
    public function getStats(): array {
        $potentialLeaks = 0;
        foreach ($this->checkouts as $info) {
            if ($info->getHeldDurationMs() > $this->config->thresholdMs) {
                $potentialLeaks++;
            }
        }
        return ['activeCheckouts' => count($this->checkouts), 'potentialLeaks' => $potentialLeaks];
    }
    
    public function checkLeaks(): void {
        foreach ($this->checkouts as $connId => $info) {
            if ($info->getHeldDurationMs() > $this->config->thresholdMs) {
                $this->logLeak($connId, $info);
            }
        }
    }
    
    private function logLeak(string $connId, CheckoutInfo $info): void {
        $msg = "POSSIBLE CONNECTION LEAK: conn={$connId}, held={$info->getHeldDurationMs()}ms";
        error_log($msg);
    }
}

File: project/drivers/driver/go/leak_detector.go
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

package scratchbird

import (
	"fmt"
	"runtime"
	"sync"
	"time"
)

// LeakDetectionConfig configures leak detection behavior
type LeakDetectionConfig struct {
	// Threshold for warning about potential leaks (default: 30 seconds)
	Threshold time.Duration
	// Whether to capture stack traces (for debugging)
	CaptureStackTrace bool
	// Check interval (default: 10 seconds)
	CheckInterval time.Duration
}

// DefaultLeakDetectionConfig returns default leak detection configuration
func DefaultLeakDetectionConfig() LeakDetectionConfig {
	return LeakDetectionConfig{
		Threshold:         30 * time.Second,
		CaptureStackTrace: false,
		CheckInterval:     10 * time.Second,
	}
}

// CheckoutInfo tracks information about a checked-out connection
type CheckoutInfo struct {
	CheckoutTime  time.Time
	GoroutineID   int64
	StackTrace    string
	Metadata      map[string]string
}

// HeldDuration returns how long the connection has been held
func (c *CheckoutInfo) HeldDuration() time.Duration {
	return time.Since(c.CheckoutTime)
}

// LeakDetector detects connection leaks
type LeakDetector struct {
	config    LeakDetectionConfig
	checkouts map[string]*CheckoutInfo
	mu        sync.RWMutex
	stopCh    chan struct{}
}

// NewLeakDetector creates a new leak detector
func NewLeakDetector(config LeakDetectionConfig) *LeakDetector {
	ld := &LeakDetector{
		config:    config,
		checkouts: make(map[string]*CheckoutInfo),
		stopCh:    make(chan struct{}),
	}
	
	// Start monitoring goroutine
	go ld.monitor()
	
	return ld
}

// Checkout registers a connection checkout
func (ld *LeakDetector) Checkout(connID string, metadata map[string]string) {
	info := &CheckoutInfo{
		CheckoutTime: time.Now(),
		GoroutineID:  getGoroutineID(),
		Metadata:     metadata,
	}
	
	if ld.config.CaptureStackTrace {
		info.StackTrace = captureStackTrace()
	}
	
	ld.mu.Lock()
	ld.checkouts[connID] = info
	ld.mu.Unlock()
}

// Checkin registers a connection return
func (ld *LeakDetector) Checkin(connID string) {
	ld.mu.Lock()
	info, exists := ld.checkouts[connID]
	if exists {
		delete(ld.checkouts, connID)
		
		// Check if held longer than threshold
		if info.HeldDuration() > ld.config.Threshold {
			// Log warning but connection was returned
			Logger.Printf("Connection %s held for %v (threshold: %v)",
				connID, info.HeldDuration(), ld.config.Threshold)
		}
	}
	ld.mu.Unlock()
}

// ActiveCheckouts returns information about active checkouts
func (ld *LeakDetector) ActiveCheckouts() map[string]*CheckoutInfo {
	ld.mu.RLock()
	defer ld.mu.RUnlock()
	
	result := make(map[string]*CheckoutInfo, len(ld.checkouts))
	for k, v := range ld.checkouts {
		// Create a copy to avoid race conditions
		infoCopy := *v
		result[k] = &infoCopy
	}
	return result
}

// ActiveCount returns the number of active checkouts
func (ld *LeakDetector) ActiveCount() int {
	ld.mu.RLock()
	defer ld.mu.RUnlock()
	return len(ld.checkouts)
}

// monitor periodically checks for potential leaks
func (ld *LeakDetector) monitor() {
	ticker := time.NewTicker(ld.config.CheckInterval)
	defer ticker.Stop()
	
	for {
		select {
		case <-ticker.C:
			ld.checkLeaks()
		case <-ld.stopCh:
			return
		}
	}
}

// checkLeaks checks for connections held beyond threshold
func (ld *LeakDetector) checkLeaks() {
	ld.mu.RLock()
	checkouts := make([]struct {
		id   string
		info *CheckoutInfo
	}, 0, len(ld.checkouts))
	
	for id, info := range ld.checkouts {
		if info.HeldDuration() > ld.config.Threshold {
			checkouts = append(checkouts, struct {
				id   string
				info *CheckoutInfo
			}{id, info})
		}
	}
	ld.mu.RUnlock()
	
	// Log warnings outside of lock
	for _, checkout := range checkouts {
		Logger.Printf("POSSIBLE CONNECTION LEAK: conn=%s held=%v goroutine=%d metadata=%v",
			checkout.id, checkout.info.HeldDuration(), checkout.info.GoroutineID, checkout.info.Metadata)
		
		if checkout.info.StackTrace != "" {
			Logger.Printf("Stack trace for connection %s:\n%s", checkout.id, checkout.info.StackTrace)
		}
	}
}

// Stop stops the leak detector
func (ld *LeakDetector) Stop() {
	close(ld.stopCh)
}

// LeakDetectionGuard provides RAII-style leak detection
type LeakDetectionGuard struct {
	connID   string
	detector *LeakDetector
	checked  bool
}

// NewLeakDetectionGuard creates a new leak detection guard
func NewLeakDetectionGuard(connID string, detector *LeakDetector, metadata map[string]string) *LeakDetectionGuard {
	detector.Checkout(connID, metadata)
	return &LeakDetectionGuard{
		connID:   connID,
		detector: detector,
		checked:  false,
	}
}

// Release marks the connection as returned
func (g *LeakDetectionGuard) Release() {
	if !g.checked {
		g.detector.Checkin(g.connID)
		g.checked = true
	}
}

// Helper to get current goroutine ID (implementation-specific)
func getGoroutineID() int64 {
	// This is a simplified version - in production use a proper implementation
	// or remove if not needed
	return 0
}

// captureStackTrace captures the current stack trace
func captureStackTrace() string {
	buf := make([]byte, 4096)
	n := runtime.Stack(buf, false)
	return string(buf[:n])
}

// LeakStats contains statistics from the leak detector
type LeakStats struct {
	ActiveCheckouts  int
	PotentialLeaks   int
}

// Stats returns current leak statistics
func (ld *LeakDetector) Stats() LeakStats {
	checkouts := ld.ActiveCheckouts()
	
	potentialLeaks := 0
	for _, info := range checkouts {
		if info.HeldDuration() > ld.config.Threshold {
			potentialLeaks++
		}
	}
	
	return LeakStats{
		ActiveCheckouts: len(checkouts),
		PotentialLeaks:  potentialLeaks,
	}
}

// Logger interface for leak detection logging
type LeakLogger interface {
	Printf(format string, v ...interface{})
}

// Default logger
var Logger LeakLogger = &defaultLogger{}

type defaultLogger struct{}

func (l *defaultLogger) Printf(format string, v ...interface{}) {
	fmt.Printf("[ScratchBird LeakDetector] "+format+"\n", v...)
}

// SetLogger sets the logger for leak detection
func SetLogger(l LeakLogger) {
	Logger = l
}
